File: workers/cpp/src/selfcheck.cpp
// The built-in case list, run by --selfcheck.
//
// The cases are the contract's edge rules, not a regression suite for this
// implementation: unclosed tag, entity without a semicolon, numeric entity,
// zero-width characters, full-width ASCII, a CJK string that must pass through
// untouched, an idempotency pair, empty input. The expected values come from
// reading docs/WORKERS.md sections 2-4 (and the parent contract delta), not from
// running this binary - otherwise the check would only prove that the code does
// what it already does.
//
// The report goes to stderr, matching the JavaScript reference and section 1's
// rule that stdout carries protocol messages and nothing else. Every line is
// pure ASCII (values are escaped for display), so the report cannot be mangled
// by a console code page that is not UTF-8.
#include "selfcheck.h"

#include <cstdio>
#include <string>
#include <vector>

#include "json.h"
#include "normalize.h"
#include "render.h"
#include "stdio_io.h"
#include "utf8.h"

#ifdef _WIN32
#include <fcntl.h>
#include <io.h>
#else
#include <unistd.h>
#endif

namespace vml {
namespace {

// Runs the protocol I/O the way the host does: over a pipe whose write end
// stays open for the whole run. Both of the failure modes below are invisible
// when a request is piped in from a file (stdin reaches EOF immediately, which
// flushes everything); against a real host they are a timeout with no error
// message at all.
bool OpenPipe(std::FILE** reader, std::FILE** writer) {
  int descriptors[2];
#ifdef _WIN32
  if (_pipe(descriptors, 4096, _O_BINARY) != 0) return false;
  *reader = _fdopen(descriptors[0], "rb");
  *writer = _fdopen(descriptors[1], "wb");
#else
  if (pipe(descriptors) != 0) return false;
  *reader = fdopen(descriptors[0], "rb");
  *writer = fdopen(descriptors[1], "wb");
#endif
  return *reader != nullptr && *writer != nullptr;
}

struct Case {
  std::string name;
  std::string actual;
  std::string expected;
};

// Renders a string as printable ASCII for the report: control characters get
// their short escapes, everything above U+007F becomes \uXXXX.
std::string EscapeForDisplay(const std::string& s) {
  static const char* kHex = "0123456789abcdef";
  std::string out;
  size_t i = 0;
  while (i < s.size()) {
    const uint32_t cp = Utf8Next(s, i);
    if (IsInvalidByteCodePoint(cp)) {
      const unsigned char b = InvalidByteFromCodePoint(cp);
      out += "\\x";
      out.push_back(kHex[(b >> 4) & 0xF]);
      out.push_back(kHex[b & 0xF]);
      continue;
    }
    switch (cp) {
      case '\n': out += "\\n"; continue;
      case '\r': out += "\\r"; continue;
      case '\t': out += "\\t"; continue;
      default: break;
    }
    if (cp >= 0x20u && cp < 0x7Fu) {
      out.push_back(static_cast<char>(cp));
      continue;
    }
    out += "\\u";
    out.push_back(kHex[(cp >> 12) & 0xF]);
    out.push_back(kHex[(cp >> 8) & 0xF]);
    out.push_back(kHex[(cp >> 4) & 0xF]);
    out.push_back(kHex[cp & 0xF]);
  }
  return out;
}

std::string NormalizedJson(const std::string& text) {
  return JsonSerialize(RenderNormalize(text));
}

std::string ExtractJson(const std::string& html) {
  return JsonSerialize(RenderExtract(html));
}

std::string FingerprintJson(const std::string& text) {
  return JsonSerialize(RenderFingerprint(text));
}

}  // namespace

int RunSelfcheck() {
  std::vector<Case> cases;
  const auto add = [&cases](const char* name, const std::string& actual,
                            const std::string& expected) {
    cases.push_back(Case{name, actual, expected});
  };

  // ---- text.normalize (section 2) ----
  add("normalize: empty input produces an empty string", NormalizedJson(""), "{\"text\":\"\"}");

  // Built from code points so that the control characters are unambiguous.
  std::string controls;
  controls += 'a';
  controls += '\0';
  controls += 'b';
  controls += '\b';
  controls += 'c';
  controls += Utf8FromCodePoint(0x200Bu);  // zero width space
  controls += 'd';
  controls += Utf8FromCodePoint(0x200Eu);  // left-to-right mark
  controls += 'f';
  controls += Utf8FromCodePoint(0x2060u);  // word joiner
  controls += 'g';
  controls += Utf8FromCodePoint(0xFEFFu);  // byte order mark
  controls += 'h';
  add("normalize: control, zero-width and BOM code points are deleted", NormalizedJson(controls),
      "{\"text\":\"abcdfgh\"}");

  add("normalize: full-width ASCII maps to ASCII",
      NormalizedJson(Utf8FromCodePoint(0xFF23u) + Utf8FromCodePoint(0xFF41u) +
                     Utf8FromCodePoint(0xFF46u) + Utf8FromCodePoint(0x00E9u)),
      "{\"text\":\"cafe\"}");

  const std::string probeA = Utf8FromCodePoint(0xFF23u) + Utf8FromCodePoint(0xFF41u) +
                             Utf8FromCodePoint(0xFF46u) + Utf8FromCodePoint(0x00E9u) +
                             Utf8FromCodePoint(0x3000u) + Utf8FromCodePoint(0x2014u) +
                             Utf8FromCodePoint(0x3000u) + "L" + Utf8FromCodePoint(0x2019u) +
                             Utf8FromCodePoint(0x00C9u) + "T" + Utf8FromCodePoint(0x00C9u) +
                             Utf8FromCodePoint(0x200Bu);
  add("normalize: full-width, ideographic space, em dash, curly quote and zero width space",
      NormalizedJson(probeA), "{\"text\":\"cafe - l'ete\"}");

  add("normalize: a run of spaces, tabs and newlines collapses to one space",
      NormalizedJson("a \t\n b"), "{\"text\":\"a b\"}");
  add("normalize: leading and trailing spaces are trimmed", NormalizedJson("   x   "),
      "{\"text\":\"x\"}");

  std::string scripts = Utf8FromCodePoint(0x65E5u) + Utf8FromCodePoint(0x672Cu) +
                        Utf8FromCodePoint(0x8A9Eu) + Utf8FromCodePoint(0x30C6u) +
                        Utf8FromCodePoint(0x30B9u) + Utf8FromCodePoint(0x30C8u) + " " +
                        Utf8FromCodePoint(0xD55Cu) + Utf8FromCodePoint(0xAD6Du) +
                        Utf8FromCodePoint(0xC5B4u) + " " + Utf8FromCodePoint(0x41Fu) +
                        Utf8FromCodePoint(0x440u) + Utf8FromCodePoint(0x438u) +
                        Utf8FromCodePoint(0x432u) + Utf8FromCodePoint(0x435u) +
                        Utf8FromCodePoint(0x442u);
  add("normalize: Han, kana, Hangul and Cyrillic pass through unchanged",
      JsonSerialize(RenderNormalize(scripts)), "{\"text\":\"" + scripts + "\"}");
  // (Compared against the input itself: the tables say nothing about these
  // scripts, so the contract requires them to be copied through untouched.)

  const std::string sharpS = "Stra" + Utf8FromCodePoint(0x00DFu) + "e " +
                             Utf8FromCodePoint(0x00C6u) + Utf8FromCodePoint(0x00D8u);
  add("normalize: lowercasing then folding (sharp s, AE, O slash)", NormalizedJson(sharpS),
      "{\"text\":\"strasse aeo\"}");

  add("normalize: ellipsis maps to three dots", NormalizedJson(std::string("A") +
      Utf8FromCodePoint(0x2026u) + "B"), "{\"text\":\"a...b\"}");

  const std::string probeB = Utf8FromCodePoint(0x0141u) + Utf8FromCodePoint(0x00F3u) + "d" +
                             Utf8FromCodePoint(0x017Au) + "   " + Utf8FromCodePoint(0x017Bu) +
                             Utf8FromCodePoint(0x00D3u) + Utf8FromCodePoint(0x0141u) +
                             Utf8FromCodePoint(0x0106u);
  add("normalize: Polish diacritics fold, inner runs collapse", NormalizedJson(probeB),
      "{\"text\":\"lodz zolc\"}");

  add("normalize: a combining mark is deleted (decomposed == composed)",
      NormalizedJson(std::string("e") + Utf8FromCodePoint(0x0301u)), "{\"text\":\"e\"}");

  const std::string dottedI = Utf8FromCodePoint(0x0130u) + "stanbul";
  add("normalize: U+0130 is absent from the lower table and folds to i, as the fold table says",
      NormalizedJson(dottedI), "{\"text\":\"istanbul\"}");

  const std::string idempotencyInput = "  " + Utf8FromCodePoint(0x00C9u) + "T" +
                                       Utf8FromCodePoint(0x00C9u) + Utf8FromCodePoint(0x2019u) +
                                       "s  " + Utf8FromCodePoint(0x2014u) + "  " +
                                       Utf8FromCodePoint(0x00DFu) + "  ";
  add("normalize: idempotency pair - normalize(normalize(x)) == normalize(x)",
      JsonSerialize(RenderNormalize(NormalizeText(NormalizeText(idempotencyInput)))),
      "{\"text\":\"ete's - ss\"}");

  // ---- text.extract (section 3) ----
  add("extract: block tags become newlines and a link keeps its text unnormalized",
      ExtractJson("<p>Hello <b>world</b></p><a href=\"/x\">Link</a> <img src=a>"),
      "{\"title\":\"\",\"text\":\"\\nHello world\\nLink \",\"links\":[{\"href\":\"/x\","
      "\"absolute\":false,\"text\":\"Link\"}],\"images\":1}");

  add("extract: an unclosed tag at end of input is dropped as a tag",
      ExtractJson("<b>bold<i"), "{\"title\":\"\",\"text\":\"bold\",\"links\":[],\"images\":0}");

  add("extract: an entity without a semicolon decodes, &nbsp; is U+00A0",
      ExtractJson("a&amp b&lt c&nbsp d"),
      "{\"title\":\"\",\"text\":\"a& b< c\xC2\xA0 d\",\"links\":[],\"images\":0}");

  add("extract: numeric entities decode (decimal, lower and upper hex)", ExtractJson("&#65;&#x42;&#X43;"),
      "{\"title\":\"\",\"text\":\"ABC\",\"links\":[],\"images\":0}");

  add("extract: script is removed with its content", ExtractJson("<p>a</p><script>var x=1;</script><p>b</p>"),
      "{\"title\":\"\",\"text\":\"\\na\\n\\nb\\n\",\"links\":[],\"images\":0}");

  add("extract: a comment is removed entirely", ExtractJson("a<!-- x -->b"),
      "{\"title\":\"\",\"text\":\"ab\",\"links\":[],\"images\":0}");

  add("extract: the title is decoded but not normalized and is not part of text",
      ExtractJson("<title>Hi &amp; Bye</title><a href=\"HTTP://X/\">Link</a>"),
      "{\"title\":\"Hi & Bye\",\"text\":\"Link\",\"links\":[{\"href\":\"HTTP://X/\","
      "\"absolute\":true,\"text\":\"Link\"}],\"images\":0}");

  add("extract: images are counted with a tag boundary respected",
      ExtractJson("<img src=a><IMG src=b><image src=c></img>"),
      "{\"title\":\"\",\"text\":\"\",\"links\":[],\"images\":2}");

  add("extract: empty input produces empty fields, never null", ExtractJson(""),
      "{\"title\":\"\",\"text\":\"\",\"links\":[],\"images\":0}");

  add("extract: a removed element with a missing closing tag runs to end of input",
      ExtractJson("<p>a</p><script>var x=1;"),
      "{\"title\":\"\",\"text\":\"\\na\\n\",\"links\":[],\"images\":0}");

  add("extract: an unknown named entity and a lone & stay verbatim", ExtractJson("&unknown; &"),
      "{\"title\":\"\",\"text\":\"&unknown; &\",\"links\":[],\"images\":0}");

  add("extract: no backtracking - &copy2024 stays literal", ExtractJson("&copy2024"),
      "{\"title\":\"\",\"text\":\"&copy2024\",\"links\":[],\"images\":0}");

  add("extract: an anchor opened while another is open closes the outer (browser rule)",
      ExtractJson("<a href=\"/one\">one<a href=\"/two\">two</a>"),
      "{\"title\":\"\",\"text\":\"onetwo\",\"links\":[{\"href\":\"/one\",\"absolute\":false,"
      "\"text\":\"one\"},{\"href\":\"/two\",\"absolute\":false,\"text\":\"two\"}],\"images\":0}");

  add("extract: end of input inside a tag drops the incomplete tag", ExtractJson("<p>abc<b"),
      "{\"title\":\"\",\"text\":\"\\nabc\",\"links\":[],\"images\":0}");

  add("extract: a lone < at end of input is literal text", ExtractJson("a<"),
      "{\"title\":\"\",\"text\":\"a<\",\"links\":[],\"images\":0}");

  add("extract: an incomplete tag contributes nothing at end of input, neither newline nor image",
      ExtractJson("<p<img"), "{\"title\":\"\",\"text\":\"\",\"links\":[],\"images\":0}");

  add("extract: CDATA content is literal text and is not re-parsed as markup",
      ExtractJson("<![CDATA[<b>raw</b>]]>"),
      "{\"title\":\"\",\"text\":\"<b>raw</b>\",\"links\":[],\"images\":0}");

  add("extract: an unclosed CDATA section keeps everything to the end of input",
      ExtractJson("<![CDATA[unclosed"),
      "{\"title\":\"\",\"text\":\"unclosed\",\"links\":[],\"images\":0}");

  add("extract: nothing inside CDATA is interpreted, entities included",
      ExtractJson("<![CDATA[a &amp; b &#65;]]>"),
      "{\"title\":\"\",\"text\":\"a &amp; b &#65;\",\"links\":[],\"images\":0}");

  add("extract: removal wins over CDATA when the CDATA sits inside a removed element",
      ExtractJson("<p><![CDATA[<script>x</script>]]></p><script>&lt;![CDATA[y]]&gt;</script>"),
      "{\"title\":\"\",\"text\":\"\\n<script>x</script>\\n\",\"links\":[],\"images\":0}");

  add("extract: a title inside an anchor belongs to the title, not to the anchor's text",
      ExtractJson("<a href=\"/x\"><title>T</title>t</a>"),
      "{\"title\":\"T\",\"text\":\"t\",\"links\":[{\"href\":\"/x\",\"absolute\":false,"
      "\"text\":\"t\"}],\"images\":0}");

  // ---- text.fingerprint (section 4) ----
  add("fingerprint: empty text has no tokens, no shingles and an all-zero hash",
      FingerprintJson(""),
      "{\"simhash\":\"0000000000000000\",\"tokens\":0,\"shingles\":0}");

  add("fingerprint: one token is one shingle equal to the hash of that token", FingerprintJson("x"),
      "{\"simhash\":\"af63f54c86021707\",\"tokens\":1,\"shingles\":1}");

  const std::string probeD = std::string("openai gpt ") + Utf8FromCodePoint(0x5DF2u) +
                            Utf8FromCodePoint(0x7ECFu) + " " + Utf8FromCodePoint(0x5DF2u) +
                            Utf8FromCodePoint(0x7ECFu);
  add("fingerprint: a CJK run of 2 emits one bigram; 4 tokens give 2 shingles",
      FingerprintJson(probeD),
      "{\"simhash\":\"2a000d8009183078\",\"tokens\":4,\"shingles\":2}");

  add("fingerprint: a token of ASCII punctuation only emits nothing", FingerprintJson("!!! ,,, a"),
      "{\"simhash\":\"af63dc4c8601ec8c\",\"tokens\":1,\"shingles\":1}");

  add("fingerprint: three tokens are one shingle", FingerprintJson("a b c"),
      "{\"simhash\":\"69cf480885ad45af\",\"tokens\":3,\"shingles\":1}");

  add("fingerprint: a CJK run of 1 emits its code point, mixed runs split",
      FingerprintJson(std::string("a") + Utf8FromCodePoint(0x4E2Du) + "b"),
      "{\"simhash\":\"27e4e607b4097edd\",\"tokens\":3,\"shingles\":1}");

  add("fingerprint: leading and trailing ASCII punctuation is trimmed from a token",
      FingerprintJson("hello, hello!"),
      "{\"simhash\":\"f76746f79c7033ff\",\"tokens\":2,\"shingles\":1}");

  // ---- protocol byte I/O (section 1) ----
  {
    std::FILE* reader = nullptr;
    std::FILE* writer = nullptr;
    std::string actual = "<could not create a pipe>";
    if (OpenPipe(&reader, &writer)) {
      const std::string request = "{\"id\":1,\"op\":\"describe\"}";
      // The write end deliberately stays open: that is what a host does.
      WriteLine(writer, request);
      LineReader pipeReader(reader);
      std::string line;
      actual = pipeReader.Next(&line) ? line
                                      : std::string("<nothing: the reader waited for end of input>");
      std::fclose(reader);
      std::fclose(writer);
    }
    add("protocol: a request line is read while the host keeps the pipe open", actual,
        "{\"id\":1,\"op\":\"describe\"}");
  }
  {
    std::FILE* reader = nullptr;
    std::FILE* writer = nullptr;
    std::string actual = "<could not create a pipe>";
    if (OpenPipe(&reader, &writer)) {
      const std::string response = "{\"id\":1,\"ok\":true,\"output\":{\"text\":\"x\"}}";
      WriteLine(writer, response);  // must be on the pipe before the host looks
      LineReader pipeReader(reader);
      std::string line;
      actual = pipeReader.Next(&line) ? line
                                      : std::string("<nothing: the response never left its buffer>");
      std::fclose(reader);
      std::fclose(writer);
    }
    add("protocol: a response line is flushed before the host reads it", actual,
        "{\"id\":1,\"ok\":true,\"output\":{\"text\":\"x\"}}");
  }
  {
    // Section 1 pins the shutdown answer as the bare envelope.
    add("protocol: shutdown answers the bare envelope", JsonSerialize(RenderShutdownAck(JsonInt(7))),
        "{\"id\":7,\"ok\":true}");
  }

  int passed = 0;
  for (const Case& c : cases) {
    const bool ok = c.actual == c.expected;
    if (ok) ++passed;
    if (ok) {
      std::fprintf(stderr, "  [ok]   %s\n", c.name.c_str());
    } else {
      std::fprintf(stderr, "  [FAIL] %s\n         expected %s\n         actual   %s\n",
                   c.name.c_str(), EscapeForDisplay(c.expected).c_str(),
                   EscapeForDisplay(c.actual).c_str());
    }
  }
  std::fprintf(stderr, "%d/%d checks passed\n", passed, static_cast<int>(cases.size()));
  std::fflush(stderr);
  return passed == static_cast<int>(cases.size()) ? 0 : 1;
}

}  // namespace vml

File: workers/go/selfcheck.go
package main

// selfcheck.go: the built-in case list behind --selfcheck (docs/WORKERS.md section 1.1).
//
// One English line per case, then "N/M checks passed", exit non-zero on failure. No protocol traffic
// on stdout in this mode: these lines are the whole stdout. The failing case prints the expected and
// the actual JSON so a divergence can be read without a debugger.
//
// The expectations here are written out as literal JSON strings (not recomputed from the code) so
// that changing a rule without changing a case fails the case.

import (
	"encoding/json"
	"fmt"
	"io"
)

type checkCase struct {
	name       string
	capability string
	// input is the protocol "input" object, as the raw JSON the host would send.
	input string
	// want is the expected output object as a JSON string; "" means "use wantFn".
	want string
	// wantFn, when set, computes the expectation from the actual output (idempotency).
	wantFn func(actual string) (string, error)
}

func checkCases() []checkCase {
	return []checkCase{
		{
			name:       "normalize/empty input stays empty",
			capability: capNormalize,
			input:      `{"text":""}`,
			want:       `{"text":""}`,
		},
		{
			name:       "normalize/full-width ASCII, ideographic space, em dash, apostrophe, ellipsis",
			capability: capNormalize,
			input:      "{\"text\":\"\uFF23\uFF41\uFF46\uFF45\u3000\u2014\u3000L\u2019\uFF25T\uFF25\u2026\"}",
			want:       `{"text":"cafe - l'ete..."}`,
		},
		{
			name:       "normalize/steps compose: E-acute is lowercased then folded",
			capability: capNormalize,
			input:      "{\"text\":\"L'\u00C9T\u00C9 \u00C9\"}",
			want:       `{"text":"l'ete e"}`,
		},
		{
			name:       "normalize/zero-width characters are deleted",
			capability: capNormalize,
			input:      "{\"text\":\"a\u200Bb\u200Dc\uFEFFd\"}",
			want:       `{"text":"abcd"}`,
		},
		{
			name:       "normalize/combining marks are deleted, so NFD equals NFC",
			capability: capNormalize,
			input:      "{\"text\":\"e\u0301 \u00E9\"}",
			want:       `{"text":"e e"}`,
		},
		{
			name:       "normalize/tab, LF and CR collapse to a single space",
			capability: capNormalize,
			input:      `{"text":"a\t\tb\n\nc\r\nd   e"}`,
			want:       `{"text":"a b c d e"}`,
		},
		{
			name:       "normalize/fold table applies after lowercasing (L-stroke, O-acute, Z-dot)",
			capability: capNormalize,
			input:      "{\"text\":\"\u0141\u00F3d\u017A   \u017B\u00D3\u0141\u0106\"}",
			want:       `{"text":"lodz zolc"}`,
		},
		{
			name:       "normalize/U+0130 folds through the fold table to i",
			capability: capNormalize,
			input:      "{\"text\":\"A\u0130B\"}",
			want:       `{"text":"aib"}`,
		},
		{
			name:       "normalize/Han, kana and Cyrillic pass through untouched",
			capability: capNormalize,
			input:      "{\"text\":\"\u6F22\u5B57\u3068\u3072\u3089\u304C\u306A \u041F\u0440\u0438\u0432\u0435\u0442\"}",
			want:       "{\"text\":\"\u6F22\u5B57\u3068\u3072\u3089\u304C\u306A \u041F\u0440\u0438\u0432\u0435\u0442\"}",
		},
		{
			name:       "normalize/idempotency: two-character folds must not re-enter the table",
			capability: capNormalize,
			input:      "{\"text\":\"\u00DF\u00C6\u00D8\u00DE\"}",
			wantFn:     idempotencyExpectation,
		},
		{
			name:       "normalize/idempotency: already-normalized text is unchanged",
			capability: capNormalize,
			input:      `{"text":"a b c"}`,
			wantFn:     idempotencyExpectation,
		},
		{
			name:       "extract/empty input yields the empty document",
			capability: capExtract,
			input:      `{"html":"","baseUrl":null}`,
			want:       `{"title":"","text":"","links":[],"images":0}`,
		},
		{
			name:       "extract/unclosed tag at end of input is dropped as a tag, newline included",
			capability: capExtract,
			input:      `{"html":"<p>a</p><div class=\"x\"","baseUrl":null}`,
			want:       `{"title":"","text":"\na\n","links":[],"images":0}`,
		},
		{
			name:       "extract/entities with and without a semicolon, numeric entities, no backtracking",
			capability: capExtract,
			input:      `{"html":"<p>&amp; &amp &lt &#65; &#x42; &#X43; &copy2024 &ampersand &unknown; &nbsp;</p>","baseUrl":null}`,
			want:       "{\"title\":\"\",\"text\":\"\\n& & < A B C &copy2024 &ampersand &unknown; \u00A0\\n\",\"links\":[],\"images\":0}",
		},
		{
			name:       "extract/removed elements, comment, CDATA, title, links, images",
			capability: capExtract,
			input:      `{"html":"<script>var x=1<\/script><style>p{}</style><!-- c --><title>T &amp; t</title><p>Hi <b>there</b></p><a href=\"nohref\">In</a><a href=\"/local\">In <i>link</i></a><a href=\"https://x.example/a\">Out</a><img src=a><img/>","baseUrl":null}`,
			want:       `{"title":"T & t","text":"\nHi there\nInIn linkOut","links":[{"href":"nohref","absolute":false,"text":"In"},{"href":"/local","absolute":false,"text":"In link"},{"href":"https://x.example/a","absolute":true,"text":"Out"}],"images":2}`,
		},
		{
			name:       "extract/quote rules: '>' inside an attribute does not end the tag, '<' of '<3' is text",
			capability: capExtract,
			input:      `{"html":"<a href=\"/a>b\">x</a> 2<3","baseUrl":null}`,
			want:       `{"title":"","text":"x 2<3","links":[{"href":"/a>b","absolute":false,"text":"x"}],"images":0}`,
		},
		{
			name:       "extract/nested anchors: the outer is reported at the inner tag, the inner becomes the open one",
			capability: capExtract,
			input:      `{"html":"<a href=\"/one\">one<a href=\"/two\">two","baseUrl":null}`,
			want:       `{"title":"","text":"onetwo","links":[{"href":"/one","absolute":false,"text":"one"},{"href":"/two","absolute":false,"text":"two"}],"images":0}`,
		},
		{
			name:       "extract/unclosed tag exactly at eof and a lone '<' at eof",
			capability: capExtract,
			input:      `{"html":"abc<b","baseUrl":null}`,
			want:       `{"title":"","text":"abc","links":[],"images":0}`,
		},
		{
			name:       "extract/a lone '<' at eof is literal text",
			capability: capExtract,
			input:      `{"html":"a<","baseUrl":null}`,
			want:       `{"title":"","text":"a<","links":[],"images":0}`,
		},
		{
			name:       "extract/a tag name alone at eof contributes no newline",
			capability: capExtract,
			input:      `{"html":"<p","baseUrl":null}`,
			want:       `{"title":"","text":"","links":[],"images":0}`,
		},
		{
			name:       "extract/an unclosed tag whose last '>' belongs to a quoted value is still unclosed (fuzz seed 7 case 2)",
			capability: capExtract,
			input:      `{"html":"<br><p title=\"unclosed&#X41</div>","baseUrl":null}`,
			want:       `{"title":"","text":"\n","links":[],"images":0}`,
		},
		{
			name:       "extract/a CDATA body inside an anchor belongs to the link text too (fuzz seed 7 case 1)",
			capability: capExtract,
			input:      `{"html":"<br><a href=\"x\"><img src=\"i.png\" alt=\"t\"><br/><![CDATA[<b>raw</b>]]></a>debut\u01c4debut< p><script src=\"a>b\"><br/>","baseUrl":null}`,
			want:       `{"title":"","text":"\n\n<b>raw</b>debutǄdebut< p>","links":[{"href":"x","absolute":false,"text":"\n<b>raw</b>"}],"images":1}`,
		},
		{
			name:       "extract/a CDATA body inside an anchor reaches the body text and the link text alike",
			capability: capExtract,
			input:      `{"html":"<a href=\"/x\"><![CDATA[<b>raw</b>]]></a>","baseUrl":null}`,
			want:       `{"title":"","text":"<b>raw</b>","links":[{"href":"/x","absolute":false,"text":"<b>raw</b>"}],"images":0}`,
		},
		{
			name:       "extract/CDATA is character data: its markup is not re-parsed",
			capability: capExtract,
			input:      `{"html":"<![CDATA[<b>raw</b>]]>","baseUrl":null}`,
			want:       `{"title":"","text":"<b>raw</b>","links":[],"images":0}`,
		},
		{
			name:       "extract/an unclosed CDATA runs to the end of input",
			capability: capExtract,
			input:      `{"html":"<![CDATA[unclosed","baseUrl":null}`,
			want:       `{"title":"","text":"unclosed","links":[],"images":0}`,
		},
		{
			name:       "extract/nothing inside CDATA is interpreted, entities included",
			capability: capExtract,
			input:      `{"html":"<![CDATA[a &amp; b &#65;]]>","baseUrl":null}`,
			want:       `{"title":"","text":"a &amp; b &#65;","links":[],"images":0}`,
		},
		{
			name:       "extract/removal wins over CDATA: a real <script> goes even when its content is CDATA",
			capability: capExtract,
			input:      `{"html":"<p><![CDATA[<script>x</script>]]></p><script>&lt;![CDATA[y]]&gt;</script>","baseUrl":null}`,
			want:       `{"title":"","text":"\n<script>x</script>\n","links":[],"images":0}`,
		},
		{
			name:       "extract/a CDATA section inside a removed element goes with it",
			capability: capExtract,
			input:      `{"html":"<script><![CDATA[gone]]></script>keep","baseUrl":null}`,
			want:       `{"title":"","text":"keep","links":[],"images":0}`,
		},
		{
			name:       "extract/a title inside an anchor belongs to the title alone",
			capability: capExtract,
			input:      `{"html":"<a href=\"/x\"><title>T</title>t</a>","baseUrl":null}`,
			want:       `{"title":"T","text":"t","links":[{"href":"/x","absolute":false,"text":"t"}],"images":0}`,
		},
		{
			name:       "extract/a title inside an anchor still takes no CDATA body (title text is exclusive)",
			capability: capExtract,
			input:      `{"html":"<a href=\"/x\"><title>T<![CDATA[q]]></title>t</a>","baseUrl":null}`,
			want:       `{"title":"Tq","text":"t","links":[{"href":"/x","absolute":false,"text":"t"}],"images":0}`,
		},
		{
			name:       "extract/hex reference decodes to its own character (&#x2014; is U+2014, not '-')",
			capability: capExtract,
			input:      `{"html":"a &#x2014; b &#8212; c","baseUrl":null}`,
			want:       "{\"title\":\"\",\"text\":\"a \u2014 b \u2014 c\",\"links\":[],\"images\":0}",
		},
		{
			name:       "fingerprint/CJK bigrams plus a Latin run",
			capability: capFingerprint,
			input:      "{\"text\":\"openai gpt \u5DF2\u7ECF \u5DF2\u7ECF\"}",
			want:       `{"simhash":"2a000d8009183078","tokens":4,"shingles":2}`,
		},
		{
			name:       "fingerprint/a single token is the single shingle and hashes alone",
			capability: capFingerprint,
			input:      `{"text":"x"}`,
			want:       `{"simhash":"af63f54c86021707","tokens":1,"shingles":1}`,
		},
		{
			name:       "fingerprint/empty text has no tokens and no shingles",
			capability: capFingerprint,
			input:      `{"text":""}`,
			want:       `{"simhash":"0000000000000000","tokens":0,"shingles":0}`,
		},
		{
			name:       "fingerprint/edge punctuation is stripped from tokens",
			capability: capFingerprint,
			input:      `{"text":"hello, hello... (world)"}`,
			wantFn:     sameAsBareWords,
		},
		{
			name:       "fingerprint/punctuation-only tokens emit nothing",
			capability: capFingerprint,
			input:      `{"text":"... !!! a"}`,
			want:       `{"simhash":"af63dc4c8601ec8c","tokens":1,"shingles":1}`,
		},
		{
			name:       "cross-capability/NFC and NFD forms of the same text normalize identically",
			capability: capNormalize,
			input:      "{\"text\":\"\u0130stanbul caf\u00E9 NFD-e\u0301\"}",
			want:       `{"text":"istanbul cafe nfd-e"}`,
		},
	}
}

// idempotencyExpectation asserts normalize(normalize(x)) == normalize(x): the second normalization
// must equal the first byte for byte, which is the contract's idempotency invariant.
func idempotencyExpectation(actual string) (string, error) {
	var first struct {
		Text string `json:"text"`
	}
	if err := json.Unmarshal([]byte(actual), &first); err != nil {
		return "", fmt.Errorf("output is not a JSON object: %v", err)
	}
	second, err := invokeCapability(capNormalize, json.RawMessage("{\"text\":"+mustJSONString(first.Text)+"}"))
	if err != nil {
		return "", fmt.Errorf("second normalization failed: %v", err)
	}
	if string(second) != actual {
		return "", fmt.Errorf("not idempotent: first=%s second=%s", actual, second)
	}
	return actual, nil
}

// sameAsBareWords asserts that stripping edge punctuation did not change the fingerprint of the
// words themselves: "hello, hello... (world)" must fingerprint exactly like "hello hello world".
func sameAsBareWords(actual string) (string, error) {
	bare, err := invokeCapability(capFingerprint, json.RawMessage(`{"text":"hello hello world"}`))
	if err != nil {
		return "", fmt.Errorf("bare-words fingerprint failed: %v", err)
	}
	var got, want fingerprintOutput
	if err := json.Unmarshal([]byte(actual), &got); err != nil {
		return "", fmt.Errorf("output is not a fingerprint object: %v", err)
	}
	if err := json.Unmarshal(bare, &want); err != nil {
		return "", fmt.Errorf("bare-words output is not a fingerprint object: %v", err)
	}
	if got != want {
		return "", fmt.Errorf("punctuation changed the fingerprint: %s vs %s", actual, bare)
	}
	return actual, nil
}

func mustJSONString(s string) string {
	encoded, err := json.Marshal(s)
	if err != nil {
		panic(err)
	}
	return string(encoded)
}

// runSelfCheck returns true when every case passed. It writes one English line per case plus the
// summary, and never writes anything to the protocol stream.
func runSelfCheck(stdout, stderr io.Writer) bool {
	cases := checkCases()
	passed := 0
	for _, c := range cases {
		actual, err := invokeCapability(c.capability, json.RawMessage(c.input))
		if err != nil {
			fmt.Fprintf(stdout, "FAIL %s: %v\n", c.name, err)
			continue
		}
		if c.wantFn != nil {
			if _, err := c.wantFn(string(actual)); err != nil {
				fmt.Fprintf(stdout, "FAIL %s: %v\n", c.name, err)
				continue
			}
			passed++
			fmt.Fprintf(stdout, "ok   %s\n", c.name)
			continue
		}
		if string(actual) != c.want {
			fmt.Fprintf(stdout, "FAIL %s\n", c.name)
			fmt.Fprintf(stdout, "       expected: %s\n", c.want)
			fmt.Fprintf(stdout, "       actual:   %s\n", string(actual))
			continue
		}
		passed++
		fmt.Fprintf(stdout, "ok   %s\n", c.name)
	}
	fmt.Fprintf(stdout, "%d/%d checks passed\n", passed, len(cases))
	if passed != len(cases) {
		fmt.Fprintf(stderr, "selfcheck failed: %d of %d cases did not pass\n", len(cases)-passed, len(cases))
		return false
	}
	return true
}
